apps/module-presale/contracts/Presale.sol

```diff
@@ -0,0 +1,336 @@
+    string private constant ERROR_INVALID_BENEFICIARY_ADDRESS    = "PRESALE_INVALID_BENEFICIARY_ADDRESS";
+    string private constant ERROR_EXCEEDS_FUNDING_GOAL           = "PRESALE_EXCEEDS_FUNDING_GOAL";
+
+    /*
+     * Roles
+     */
+
+    bytes32 public constant START_ROLE = keccak256("START_ROLE");
+    bytes32 public constant BUY_ROLE   = keccak256("BUY_ROLE");
+
+    /*
+     * Properties
+     */
+
+    // Token that will be accepted as payment for
+    // purchasing project tokens.
+    ERC20 public daiToken;
+
+    // Token that is being offered for purchase in the sale.
+    MiniMeToken public projectToken;
+    TokenManager public projectTokenManager;
+
+    // Funding goal and total dai raised in the sale.
+    // Note: no further purchases will be allowed after daiFundingGoal is reached.
+    uint256 public daiFundingGoal;
+    uint256 public totalDaiRaised;
+
+    // Percentage of the total supply of project tokens that will be offered to contributors,
+    // in this pre-sale and in further fundraising stages.
+    uint256 public percentSupplyOffered; // Represented in PPM, see below
+
+    // Once the sale is closed, totalDaiRaised is split according to
+    // percentFundingForBeneficiary, between beneficiaryAddress and fundraisingPool.
+    address public fundraisingPool;
+    address public beneficiaryAddress;
+    uint256 public percentFundingForBeneficiary; // Represented in PPM, see below
+
+    // Date when the sale is started and its state is Funding.
+    uint64 public startDate;
+
+    // Vesting parameters.
+    // Note: startDate also represents the starting date for all vested project tokens.
+    uint64 public vestingCliffPeriod;
+    uint64 public vestingCompletePeriod;
+
+    // Period after startDate, in which the sale is Funding and accepts contributions.
+    // If the daiFundingGoal is not reached within it, the sale cannot be Closed
+    // and the state switches to Refunding, allowing dai refunds.
+    uint64 public fundingPeriod;
+
+    // Number of project tokens that will be sold for each dai.
+    // Calculated after initialization from the values CONNECTOR_WEIGHT_PPM and percentSupplyOffered.
+    uint256 public daiToProjectTokenExchangeRate;
+
+    // Percentages are represented in the PPM range (Parts per Million) [0, 1000000]
+    // 25% => 0.25 * 1e6
+    // 50% => 0.50 * 1e6
+    uint256 public constant PPM = 1000000;
+
+    // Used to calculate daiToProjectTokenExchangeRate.
+    uint256 public constant CONNECTOR_WEIGHT_PPM = 100000; // 10%
+
+    // Keeps track of how much dai is spent, per purchase, per buyer.
+    // This is used when refunding purchases.
+    mapping(address => mapping(uint256 => uint256)) public purchases;
+    /*      |                  |          |
+     *      |                  |          daiSpent
+     *      |                  purchaseId
+     *      buyer
+     */
+
+    bool private saleClosed;
```

Comment:
Seems unnecessary to save this in storage

Reply:
I know that it looks a little awkward, but notice how it is actually needed for the close() function. Without it, it'd be difficult to differentiate the GoalReached and the Closed states. An alternative would be to re-use one of the other state variables for this, since they are not needed anymore; it would be cheaper, but less explicit.

Reply:
Ah I see; originally I thought closed and goal reached could be determined just by the time / amount bought but I see what you mean.